src/ast.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 VMware, Inc.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+use serde_json::Value;
+
+/// A path is a tree of selector nodes.
+///
+/// For example, the JSONPath `$.foo.bar` yields this AST:
+///
+/// ```text
+///              ^
+///             / \
+///            ^   \___ DotName("bar")
+///           / \
+/// Root ___ /   \___ DotName("foo")
+/// ```
+///
+/// A more complicated example: `$.foo[1,2]["bar"]`:
+///
+/// ```text
+///                 ^
+///                / \
+///               ^   \___ Union
+///              / \            \
+///             /   \___ Union   \___ [Field("bar")]
+///            /              \
+///           ^                \___ [Number(1), Number(2)]
+///          / \
+/// Root ___/   \___ DotName("foo")
+/// ```
+///
+/// Selectors are left associative, thus `$.foo[1,2]["bar"]` behaves
+/// like (pseudocode) `(($.foo)[1,2])["bar"]`; thus the root of the resulting
```

Comment:
Is it worth pointing out that the `Root` selector (`$`) is not related to the root of the AST? Might confuse people otherwise.